### PR TITLE
Add Support for Unitful contours

### PR DESCRIPTION
**Pull request by dstansby**
This is extracted from https://github.com/sunpy/sunpy/pull/4875 as a minimal change to get contours working. We can add the example from that PR later when this has gone in with tests and a changelog.

### CODE REVIEW

File: sunpy/map/tests/test_mapbase.py [hunk 1]
```diff
@@ -1073,16 +1074,25 @@ def test_contour(simple_map):
...
     contours_cm = simple_map.contour(150 * u.cm)
     for c1, c2 in zip(contours, contours_cm):
         np.all(c1 == c2)
 
-    with pytest.raises(u.UnitsError, match='level must be an astropy quantity convertible to m'):
+    # Percentage
+    contours_percent = simple_map.contour(100 * u.percent)
+    contours_ref = simple_map.contour(np.max(simple_map.data) * simple_map.unit)
+    for c1, c2 in zip(contours_percent, contours_ref):
+        np.all(c1 == c2)
```
Comment (dpshelio) [inline sunpy/map/tests/test_mapbase.py:1091]: Should this have an assert?
```suggestion
        assert np.all(c1 == c2)
```

Comment (dstansby) [reply sunpy/map/tests/test_mapbase.py]: 🙈 good catch!

File: sunpy/map/mapbase.py [hunk 2]
```diff
@@ -2137,16 +2137,50 @@ def draw_rectangle(self, bottom_left, *, top_right=None, width: u.deg = None, he
...
+        ``warn_units`` allows legacy option of passing levels that aren't Quantities.
+        This can be removed in sunpy 3.1.
+        """
+        levels = np.atleast_1d(levels)
+        if not hasattr(levels, 'unit'):
+            if self.unit is None:
+                # No map units, so allow non-quantity through
+                return levels
+            elif warn_units:
+                warnings.warn('Passing contour levels without units is deprecated. '
```
Comment (hayesla) [inline sunpy/map/mapbase.py:2153]: will percentage inputs be deprecated? if not this warning really should also say to either pass levels with units or as a percentage for when someone passes a list of numbers meaning to be percentages.

File: sunpy/map/mapbase.py [hunk 3]
```diff
@@ -2137,16 +2137,50 @@ def draw_rectangle(self, bottom_left, *, top_right=None, width: u.deg = None, he
...
+                              f'Pass levels in units convertible to {self.unit} to '
```
Comment (hayesla) [inline sunpy/map/mapbase.py:2154]:
```suggestion
                              f'Pass levels in units convertible to units of the map ({self.unit}) to '
```

File: sunpy/map/mapbase.py (hunk 2 above)
Comment (dstansby) [reply sunpy/map/mapbase.py:2153]: No, they're not being deprecated (or removed), I'll updating the message.

File: sunpy/map/mapbase.py (hunk 3 above)
Comment (dstansby) [reply sunpy/map/mapbase.py]: I've updated the message, let me know what you think!

**Review by hayesla — APPROVED** [review]

File: sunpy/map/mapbase.py [hunk 4]
```diff
@@ -2137,16 +2137,50 @@ def draw_rectangle(self, bottom_left, *, top_right=None, width: u.deg = None, he
...
+        ``warn_units`` allows legacy option of passing levels that aren't Quantities.
+        This can be removed in sunpy 3.1.
+        """
+        levels = np.atleast_1d(levels)
+        if not hasattr(levels, 'unit'):
+            if self.unit is None:
+                # No map units, so allow non-quantity through
+                return levels
+            elif warn_units:
+                warnings.warn('Passing contour levels that are not an astropy Quantity is deprecated. '
```
Comment (Cadair) [inline sunpy/map/mapbase.py:2153]: This should have a comment or something saying that it was deprecated in 3.0.

**Review by nabobalis — APPROVED** [review]

**Pull request merged**